xcode/Subconscious/Shared/Components/Detail/UserProfileDetailView.swift

```diff
@@ -131,6 +194,121 @@ struct UserProfileDetailModel: ModelProtocol {
+                .map { following in
+                    UserProfileDetailAction.populateFollowingStatus(following)
+                }
+                .eraseToAnyPublisher()
+            
+            return Update(state: state, fx: fx)
+            
+        case .populateFollowingStatus(let following):
+            var model = state
+            model.isFollowingUser = following
+            return Update(state: model)
+            
+        // MARK: Following
+        case .requestFollow:
+            return update(state: state, action: .presentFollowSheet(true), environment: environment)
+            
+        case .attemptFollow:
+            guard let did = state.followUserSheet.followUserForm.did.validated else {
+                return Update(state: state)
+            }
+            guard let petname = state.followUserSheet.followUserForm.petname.validated else {
+                return Update(state: state)
+            }
+            
+            let fx: Fx<UserProfileDetailAction> =
+                environment.data.addressBook
+                .followUserAsync(did: did, petname: petname, preventOverwrite: true)
+                .map({ _ in
+                    UserProfileDetailAction.succeedFollow(did: did, petname: petname)
+                })
+                .catch { error in
+                    Just(UserProfileDetailAction.failFollow(error: error.localizedDescription))
+                }
+                .eraseToAnyPublisher()
+            
+            return Update(state: state, fx: fx)
+            
+        case .succeedFollow(let did, let petname):
+            return update(
+                state: state,
+                actions: [
+                    .presentFollowSheet(false),
+                    .fetchFollowingStatus(did, petname)
+                ],
+                environment: environment
+            )
+            
+        case .failFollow(error: let error):
+            var model = state
+            model.failFollowErrorMessage = error
+            return Update(state: model)
+            
+        case .dismissFailFollowError:
+            var model = state
+            model.failFollowErrorMessage = nil
+            return Update(state: model)
+            
+        // MARK: Unfollowing
+        case .presentUnfollowConfirmation(let presented):
+            var model = state
+            model.isUnfollowConfirmationPresented = presented
+            return Update(state: model)
+            
+        case .requestUnfollow:
+            return update(state: state, action: .presentUnfollowConfirmation(true), environment: environment)
+            
+        case .attemptUnfollow:
+            guard let petname = state.user?.petname, let did = state.user?.did else {
+                return Update(state: state)
+            }
+            
+            let fx: Fx<UserProfileDetailAction> =
+            environment.data.addressBook
+                .unfollowUserAsync(did: did)
```

Comment:
We unfollow by _DID_ here because the petname may not match what's in your address book.